akka-stream/src/main/scala/akka/stream/impl/LazySource.scala

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.impl
+
+import akka.stream.impl.fusing.SubSink
+import akka.stream.scaladsl.SinkQueueWithCancel
+import akka.stream.{ Attributes, Outlet, SourceShape }
+import akka.stream.scaladsl.{ Keep, Source }
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+import scala.util.{ Failure, Success, Try }
+
+object LazySource {
+  def apply[T, M](createSource: ⇒ Source[T, M]): LazySource[T, M] = new LazySource[T, M](createSource)
+}
+
+final class LazySource[T, M](createSource: ⇒ Source[T, M]) extends GraphStageWithMaterializedValue[SourceShape[T], Future[M]] {
+  val out = Outlet[T]("LazySource.out")
+  override val shape = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val matPromise = Promise[M]()
+    val logic = new GraphStageLogic(shape) with OutHandler {
+
+      var sinkQueue: SinkQueueWithCancel[T] = _
+
+      val cb = getAsyncCallback[Try[Option[T]]] {
```

Comment:
I would alternatively recommend using SubSink. You create a `SubSink` via `SubSinkInlet`, attach it to the source created by the factory method. Then you can use the usual grab/pull style coding there. 

The benefit compared to the current approach is that it would not break fusing and would be able to use the internal queue instead of the actor mailbox instead which is one magnitude faster AFAIK.